rust/arrow/src/record_batch.rs

```diff
@@ -103,6 +103,56 @@ impl RecordBatch {
         RecordBatch { schema, columns }
     }
 
+    /// Creates a new [`RecordBatch`] with no columns
+    ///
+    /// TODO add an code example using `append`
+    pub fn new() -> Self {
+        Self {
+            schema: Arc::new(Schema::empty()),
+            columns: Vec::new(),
+        }
+    }
+
+    /// Appends the `field_array` array to this `RecordBatch` as a
+    /// field named `field_name`.
+    ///
+    /// TODO: code example
+    ///
+    /// TODO: on error, can we return `Self` in some meaningful way?
+    pub fn append(self, field_name: &str, field_values: ArrayRef) -> Result<Self> {
```

Comment:
Given that `ArrayRef` is backed by an `Arc`, how about taking it by reference then cloning it?

Am I correct that if this errors, the underlying `ArrayRef` could get dropped if it's not a clone in itself?
Something like:

```rust
let array = Int32Array::from(vec![1, 2, 3]);
let array_ref = Arc::new(array);

// append to existing batch
let mut batch = ...;
// assume that this fails, does array_ref get dropped as we no longer have references to it?
batch = batch.append("ints", array_ref)?;
```

Reply:
I guess I was thinking we would leave the "do I want to have access to the `ArrayRef` after the call to `append` up to the caller, 

So like

```
// append to existing batch
let mut batch = ...;
// We want to use array_ref later, so pass in a clone
batch = batch.append("ints", Arc::clone(array_ref))?;
```

Which I think is a common pattern in the Rust libraries (that if a function needs to clone its argument, it takes it by value instead and thus gives the caller a chance to avoid an extra copy if they don't need the argument again)

But in this case with an `Arc` that is cheap to clone I don't think the difference really matters from a performance perspective